module/x/gravity/keeper/attestation_handler.go

```diff
@@ -21,317 +22,510 @@ var _ types.DistributionKeeper = (*distrkeeper.Keeper)(nil)
+	// Cosmos denoms have a system that is much more full featured, with enterprise-ready token denominations.
+	// There is a DenomUnits array that tells you what the name of each denomination of the
+	// token is.
+	// To correlate this with an ERC20 "decimals" field, we have to search through the DenomUnits array
+	// to find the DenomUnit which matches up to the main token "display" value. Then we take the
+	// "exponent" from this DenomUnit.
+	// If the correct DenomUnit is not found, it will default to 0. This will result in there being no decimal places
+	// in the token's ERC20 on Ethereum. So, for example, if this happened with Atom, 1 Atom would appear on Ethereum
+	// as 1 million Atoms, having 6 extra places before the decimal point.
+	// This will only happen with a Denom Metadata which is for all intents and purposes invalid, but I am not sure
+	// this is checked for at any other point.
+	decimals := uint32(0)
+	for _, denomUnit := range metadata.DenomUnits {
+		if denomUnit.Denom == metadata.Display {
+			decimals = denomUnit.Exponent
+			break
 		}
+	}
+
+	if decimals != uint32(claim.Decimals) {
+		return sdkerrors.Wrap(
+			types.ErrInvalid,
+			fmt.Sprintf("ERC20 decimals %d does not match denom decimals %d", claim.Decimals, decimals))
+	}
+
+	// Add to denom-erc20 mapping
+	a.keeper.setCosmosOriginatedDenomToERC20(ctx, claim.CosmosDenom, *tokenAddress)
 
-		if !invalidAddress { // valid address so far, try to lock up the coins in the requested cosmos address
-			if err := a.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, nativeReceiver, coins); err != nil {
-				// someone attempted to send tokens to a blacklisted user from Ethereum, log and send to Community pool
-				hash, _ := claim.ClaimHash()
-				a.keeper.logger(ctx).Error("Blacklisted deposit",
-					"cause", err.Error(),
-					"claim type", claim.GetType(),
-					"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
-					"nonce", fmt.Sprint(claim.GetEventNonce()),
+	ctx.EventManager().EmitTypedEvent(
+		&types.EventERC20DeployedClaim{
+			Token: tokenAddress.GetAddress().Hex(),
+			Nonce: strconv.Itoa(int(claim.GetEventNonce())),
+		},
+	)
+	return nil
+}
+
+// Upon acceptance of sufficient ValsetUpdated claims: update LastObservedValset, mint cosmos-originated relayer rewards
+// so that the reward holder can send them to cosmos
+func (a AttestationHandler) handleValsetUpdated(ctx sdk.Context, claim types.MsgValsetUpdatedClaim) error {
+	rewardAddress, err := types.NewEthAddress(claim.RewardToken)
+	if err != nil {
+		return sdkerrors.Wrap(err, "invalid reward token on claim")
+	}
+	// TODO here we should check the contents of the validator set against
+	// the store, if they differ we should take some action to indicate to the
+	// user that bridge highjacking has occurred
+	a.keeper.SetLastObservedValset(ctx, types.Valset{
+		Nonce:        claim.ValsetNonce,
+		Members:      claim.Members,
+		Height:       0,
```

Comment:
height is zero? I don't think that's right.

Reply:
this height value should be the block height the valset was produced at, so we would lookup the valset nonce in the store, then we can camopre the members list and panic if they don't match (valset not produced by consensus, bridge highjacked) and take the block value.